pkg/services/ngalert/schedule/schedule.go

```diff
@@ -431,18 +445,26 @@ func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key ngmodels.AlertR
 	for {
 		select {
 		// used by external services (API) to notify that rule is updated.
-		case lastVersion := <-updateCh:
+		case ctx := <-updateCh:
 			// sometimes it can happen when, for example, the rule evaluation took so long,
 			// and there were two concurrent messages in updateCh and evalCh, and the eval's one got processed first.
 			// therefore, at the time when message from updateCh is processed the current rule will have
 			// at least the same version (or greater) and the state created for the new version of the rule.
-			if currentRuleVersion >= int64(lastVersion) {
-				logger.Info("Skip updating rule because its current version is actual", "version", currentRuleVersion, "newVersion", lastVersion)
+			if currentRuleVersion >= int64(ctx.Version) {
+				logger.Info("Skip updating rule because its current version is actual", "version", currentRuleVersion, "newVersion", ctx.Version)
 				continue
 			}
-			logger.Info("Clearing the state of the rule because version has changed", "version", currentRuleVersion, "newVersion", lastVersion)
+
+			// we can continue directly if the rule is paused because in the next tick it will be picked up. The state
+			// will be cleaned up and the routine will be stopped and removed from the registry
+			if ctx.IsPaused {
+				logger.Info("Skip updating rule because it is paused")
+				clearState(grafanaCtx, ngmodels.StateReasonPaused)
+				continue
+			}
+			logger.Info("Clearing the state of the rule because version has changed", "version", currentRuleVersion, "newVersion", ctx.Version)
```

Comment:
let's reduce branching here.
```go
reason := ngmodels.StateReasonUpdated
if ctx.IsPaused {
    reason = ngmodels.StateReasonPaused
}
logger.Info("Clearing the state of the rule because version has changed", "version", currentRuleVersion, "newVersion", ctx.Version, "reason", reason)
clearState(grafanaCtx, reason)
```

Reply:
we can change wording a little bit and say that "Clearing the state of the rule because it was updated"